integrations/linear/src/webhooks.ts

```diff
@@ -0,0 +1,341 @@
+    return this.runTask(
+      key,
+      async (client, task) => {
+        const payload = await client.updateWebhook(params.id, params.input);
+        return withoutFunctions({
+          ...payload,
+          webhook: await payload.webhook,
+        });
+      },
+      {
+        name: "Update Webhook",
+        params,
+      }
+    );
+  }
+}
+
+type LinearEvents = (typeof events)[keyof typeof events];
+
+const TriggerParamsSchema = z.object({
+  resourceTypes: z.array(WebhookResourceTypeSchema),
+  teamId: z.string().optional(),
+  allPublicTeams: z.boolean().optional(),
+  actionTypes: z.array(WebhookActionTypeSchema).optional(),
+});
+
+export type TriggerParams = z.infer<typeof TriggerParamsSchema>;
+
+type CreateTriggersResult<TEventSpecification extends LinearEvents> = ExternalSourceTrigger<
+  TEventSpecification,
+  ReturnType<typeof createWebhookEventSource>
+>;
+
+export function createTrigger<TEventSpecification extends LinearEvents>(
+  source: ReturnType<typeof createWebhookEventSource>,
+  event: TEventSpecification,
+  params: TriggerParams
+  // options: {
+  //   actionTypes?: WebhookActionType[]; // via params
+  // }
+): CreateTriggersResult<TEventSpecification> {
+  return new ExternalSourceTrigger({
+    event,
+    params,
+    source,
+    options: {},
+  });
+}
+
+const WebhookRegistrationDataSchema = z.object({
+  success: z.literal(true),
+  webhook: z.object({
+    id: z.string(),
+    enabled: z.boolean(),
+  }),
+});
+type WebhookRegistrationData = z.infer<typeof WebhookRegistrationDataSchema>;
+
+export function createWebhookEventSource(
+  integration: Linear
+): ExternalSource<Linear, TriggerParams, "HTTP", {}> {
+  return new ExternalSource("HTTP", {
+    id: "linear.webhook",
+    schema: TriggerParamsSchema,
+    // optionSchema: z.object({ }),
+    version: "0.1.0",
+    integration,
+    // TODO: filter by actionTypes
+    // filter: (params, options) => ({ }),
+    key: (params) => `${params.teamId ? params.teamId : "all"}`,
+    handler: webhookHandler,
+    register: async (event, io, ctx) => {
+      const { params, source: httpSource, options } = event;
+
+      // (key-specific) stored data, undefined if not registered yet
+      const webhookData = WebhookRegistrationDataSchema.safeParse(httpSource.data);
+
+      console.log(webhookData);
+      if (!webhookData.success) {
+        console.log(webhookData.error);
+      }
+
+      // set of events to register
+      const allEvents = Array.from(new Set([...options.event.desired, ...options.event.missing]));
+      const registeredOptions = {
+        event: allEvents,
+      };
+
+      // easily identify webhooks on linear
+      const label = `trigger.${params.teamId ? params.teamId : "all"}`;
+
+      // TODO: remove tunnel
+      const url = process.env["DEV_TUNNEL"]
+        ? httpSource.url.replace("http://localhost:3030", process.env["DEV_TUNNEL"])
+        : httpSource.url;
```

Comment:
This has been useful for dev via ngrok et al. Are there plans to integrate tunnels for webhooks more generally? Maybe they already exist and I missed something!

Reply:
The `httpSource.url` is provided by the webapp, and uses the `APP_ORIGIN` env var that needs to be set when running the webapp locally.  For example: `APP_ORIGIN='http://abc1234.ngrok.io'`

Reply:
Excellent. Any unintended side-effects? Would rather not force local traffic via the tunnel, just incoming webhooks.